### ISSUE TITLE
Local extensions loaded twice when included as package dependency

### ISSUE THREAD

**Issue by echocrow** [directus/directus#21926]
### Describe the Bug

When packaging extensions locally and listing them as dependency in `package.json`, Directus currently will load them twice.

_Background:_
We've "installed" our local extension packages as dependencies in the Directus app package, so that our monorepo tooling (e.g. Turborepo) picks up on the extensions as dependencies, and knows to build those first before deploying Directus (e.g. to Directus Cloud).

Temp fix is to configure local Directus with e.g. `EXTENSIONS_PATH="./null"`.

### To Reproduce

`./package.json`
```jsonc
{
  // ...
  "dependencies": {
    "my-fancy-extension": "workspace:^",
    "directus": "^10.10.4",
    "pg": "^8.11.3"
  },
  // ...
}
```
`./extensions/my-fancy-extension/package.json`
```jsonc
{
  "name": "my-fancy-extension":",
  "icon": "extension",
  // ...
}
```

Then start Directus.

**Expected:**
- Only one instance of `my-fancy-extension` in the `directus_extensions` table
- `my-fancy-extension` listed only once on `/admin/settings/extensions`
- Any hooks (e.g. `init('app.after', () => ...)` registered and fired once

**Actual:**
- Two records of `my-fancy-extension` in the `directus_extensions` table
- `my-fancy-extension` listed twice on `/admin/settings/extensions`
- Hooks (e.g. `init('app.after', () => ...)` registered and fired twice

### Directus Version

v10.10.4

### Hosting Strategy

Self-Hosted (Docker Image)

**Comment by robinsonjohn** [comment]
I am also experiencing a similar issue since the release of Directus 10.10.x with the extensions marketplace.

**Comment by echocrow** [comment]
minor update, found a temp workaround for us:

instead of declaring local extensions as `dependencies` (which is currently causing Directus to load them twice), we shifted to listing them as `optionalDependencies`.

**Comment by robinsonjohn** [comment]
In my particular instance, I am using the [Sane Image Size](https://github.com/martabitbrain/directus-extension-sane-image-size) plugin, but it does not appear on the newly developed Marketplace.

Therefore, I am adding it as a dependency in my `package.json` file:

```
"dependencies": {
    "directus-extension-sane-image-size": "https://github.com/martabitbrain/directus-extension-sane-image-size.git#v1.1.1"
}
```

...then my install script runs the following:

```
cp -r ./node_modules/directus-extension-sane-image-size ./extensions
```

This allows me to keep the extension up to date with the current repository and exclude its files from my version controlled application.

It was working great until the update which included the Marketplace was released. Now, the extension loads as expected as a `local` extension, but Directus is also loading it as a `module`, but the module version appears as  "missing" (loaded twice).

![Screenshot 2024-03-28 at 9 05 25 AM](https://github.com/directus/directus/assets/24327848/dcf688ad-4921-4552-8506-9a8b658cf67c)

![Screenshot 2024-03-28 at 9 07 18 AM](https://github.com/directus/directus/assets/24327848/86cfcff0-1d7d-4a5a-9f27-da6a9975e40e)

**Comment by echocrow** [comment]
@robinsonjohn hmm, I suspect that custom remove extension not being picked up might be a different bug.

in the meantime, did you try moving `directus-extension-sane-image-size` into `devDependencies` or `optionalDependencies`? I suspect the same workaround may also work for you.

(BTW, you can probably also symlink (`ln -s`) instead of copy (`cp -r`) the extension. Not a big difference, but saves from the extra coping, and and re-copying after package updates. - We actually used to symlink all our `./extensions` into a custom `./.directus/extensions` dir ourselves before v10.10 just in order to not prefix our source foldes with `directus-extension-` 😅.)

**Comment by robinsonjohn** [comment]
@echocrow Thanks for the suggestions. I had attempted to use a symlink, but Directus completely ignored the extensions as if they weren't there. I had to copy the files themselves in order to get it to work.

I'll go back and attempt the `devDependencies` suggestion...

**Comment by br41nslug** [comment]
> I had attempted to use a symlink, but Directus completely ignored the extensions as if they weren't there.

That is to be expected if you're using docker as you cannot symlink between the container and host OS.

**Comment by piotr-cz** [comment]
For me the solution to this problem is:

- In extensions folder use same subfolder as extension scope-less name defined in `package.json` . For example:
  - package.json name is `@acme/directus-extension-foobar-display`
  - path should be `extensions/directus-extension-foobar-display`  
    previously I didn't use `directus-extension-` prefix, as package was in Directus extensions folder

- local extension must be installed before module extension

This should work as long as these code is still present:

https://github.com/directus/directus/blob/32d5650f4f4e2c4ac193e893367d6ea7569c5265/api/src/extensions/lib/get-extensions-settings.ts#L126-L138

___
Update: Actually this didn't work for me, but may without using package scope (`@acme/`)

**Comment by piotr-cz** [comment]
For my case I declared extensions folder as a project workspace (`"workspaces": ["extensions/*"]`) in package.json, but didn't add my extensions as project dependency.

On `npm install` my extension dependencies were added to project (expected), but Directus installed only the `local` source and omitted `module` source - this is what I want.

**Comment by echocrow** [comment]
> For my case I declared extensions folder as a project workspace ("workspaces": ["extensions/*"]) in package.json, but didn't add my extensions as project dependency.

we landed on the same solution.

if anyone's using monorepo tools like turborepo: we've defined our `./extensions/*` packages as `optionalDependencies` (instead of `dependencies` or `devDependencies`) to the Directus app project. this way, Directus will _not_ pick them up as package dependency (only once as local extension), while Turborepo does pick them up and will run e.g. build steps of all extensions before the build/deploy step of the main Directus app package.

**Comment by piotr-cz** [comment]
It makes sense to me now. Directus extensions may be installed as:
- [npm packages](https://directus.io/docs/self-hosting/including-extensions#npm-packages) / Project dependencies (`type: module`) 
- [Local extensions](https://directus.io/docs/self-hosting/including-extensions#directly-including-extensions) (`type:local`)
- Marketplace extensions (`type:registry`)

_Local extensions_ should not be included as project dependencies in package.json, because Directus uses own Extension manager to install them.

However _Local extensions_ may be part of project workspace, so it's possible to install their dependencies (on `npm install`) or run their build scripts (ie. `npm run build --workspace extensions`) straight from the project directory.

**Comment by bellahadid27211-lab** [comment]
Yep, that’s exactly what we ended up doing too. Declared the extensions as
a workspace but kept them out of dependencies so Directus handles them as
local extensions, while Turborepo still runs their build steps before the
main app. Makes everything work smoothly.
